whois-nrtm4-client/src/main/java/net/ripe/db/nrtm4/client/importer/DeltaImporter.java

```diff
@@ -0,0 +1,175 @@
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DeltaImporter.class);
+
+    private final NrtmRestClient nrtmRestClient;
+
+    private final Nrtm4ClientRepository nrtm4ClientRepository;
+
+    private final Nrtm4ClientInfoRepository nrtm4ClientInfoRepository;
+
+
+    public static final String RECORD_SEPARATOR = "\u001E";
+
+    public DeltaImporter(final NrtmRestClient nrtmRestClient,
+                         final Nrtm4ClientRepository nrtm4ClientRepository,
+                         final Nrtm4ClientInfoRepository nrtm4ClientInfoRepository) {
+        this.nrtmRestClient = nrtmRestClient;
+        this.nrtm4ClientRepository = nrtm4ClientRepository;
+        this.nrtm4ClientInfoRepository = nrtm4ClientInfoRepository;
+    }
+
+    @Override
+    public void doImport(final String source, final UpdateNotificationFileResponse updateNotificationFile){
+        final List<UpdateNotificationFileResponse.NrtmFileLink> deltas = getNewDeltas(source, updateNotificationFile);
+
+        if (deltas.isEmpty()) {
+            LOGGER.info("No new deltas to be processed");
+            return;
+        }
+
+        deltas.forEach(delta -> {
+            final byte[] deltaFilePayload = nrtmRestClient.getDeltaFiles(delta.getUrl());
+
+            if (deltaFilePayload == null || deltaFilePayload.length == 0){
+                LOGGER.error("This cannot happen. UNF has a non-existing delta");
+                return;
+            }
+
+            final String[] deltaFileResponse = StringUtils.split(new String(deltaFilePayload, StandardCharsets.UTF_8), RECORD_SEPARATOR);
+
+            final Metadata metadata = getMetadata(deltaFileResponse);
+
+            final List<MirrorDeltaInfo> mirrorObjectInfos = getMirrorDeltaObjects(deltaFileResponse);
+
+            if (!delta.getHash().equals(calculateSha256(deltaFilePayload))){
+                LOGGER.error("Snapshot hash doesn't match, skipping import");
+                return;
+            }
+
+            if (!metadata.sessionId().equals(updateNotificationFile.getSessionID())){
+                LOGGER.error("The session is not the same in the UNF and snapshot");
+                return;
+            }
+
+            mirrorObjectInfos.forEach(this::applyDeltaRecord);
+
+            nrtm4ClientInfoRepository.saveDeltaFileVersion(source, metadata.version, metadata.sessionId());
+        });
+    }
+
+    private void applyDeltaRecord(final MirrorDeltaInfo deltaInfo){
+        if (deltaInfo.getAction().equals(MirrorDeltaInfo.Action.ADD_MODIFY)){
+
+            final RpslObjectUpdateInfo rpslObjectUpdateInfo = nrtm4ClientRepository
+                    .getMirroredObjectId(deltaInfo.getRpslObject().getType(), deltaInfo.getRpslObject().getKey().toString());
+            if (rpslObjectUpdateInfo == null) {
+                applyDeltaCreation(deltaInfo);
+                return;
+            }
+            applyDeltaUpdate(deltaInfo, rpslObjectUpdateInfo);
```

Comment:
Where do you catch DB exceptions? Do you rollback all deltas if one fails to commit ? Consider wrapping in a `@Transactional` annotation so it's all or nothing.

Reply:
yes, I created a subtask for rolling back. It is no working property in this PR...and there are already a lot of changes